test/extensions/filters/http/ext_proc/streaming_integration_test.cc

```diff
@@ -261,6 +407,286 @@ TEST_P(StreamingIntegrationTest, GetAndProcessBufferedResponseBody) {
+       response_size](grpc::ServerReaderWriter<ProcessingResponse, ProcessingRequest>* stream) {
+        ProcessingRequest header_req;
+        if (!stream->Read(&header_req)) {
+          return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "expected message");
+        }
+        if (!header_req.has_request_headers()) {
+          return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "expected request headers");
+        }
+
+        ProcessingResponse header_resp;
+        header_resp.mutable_request_headers();
+        auto* override = header_resp.mutable_mode_override();
+        override->set_response_header_mode(ProcessingMode::SKIP);
+        override->set_response_body_mode(ProcessingMode::STREAMED);
+        stream->Write(header_resp);
+
+        ProcessingRequest body_req;
+        uint32_t total_response_size = 0;
+        Buffer::OwnedImpl allData;
+
+        do {
+          if (!stream->Read(&body_req)) {
+            return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "expected message");
+          }
+          if (!body_req.has_response_body()) {
+            return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "expected response body");
+          }
+          total_response_size += body_req.response_body().body().size();
+          // Save all the chunks in a buffer so that we can calculate a hash.
+          allData.add(body_req.response_body().body());
+
+          ProcessingResponse body_resp;
+          body_resp.mutable_response_body();
+          stream->Write(body_resp);
+        } while (!body_req.response_body().end_of_stream());
+
+        processor_response_hash_ = HashUtil::xxHash64(allData.toString());
+        if (total_response_size != response_size) {
+          return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT,
+                              absl::StrFormat("Received %u response bytes; wanted %u",
+                                              total_response_size, response_size));
+        }
+
+        return grpc::Status::OK;
+      });
+
+  initializeConfig();
+  HttpIntegrationTest::initialize();
+  Http::TestRequestHeaderMapImpl headers;
+  HttpTestUtility::addDefaultHeaders(headers);
+  headers.addCopy(LowerCaseString("response_size_bytes"), response_size);
+  sendGetRequest(headers);
+
+  ASSERT_TRUE(client_response_->waitForEndStream());
+  EXPECT_TRUE(client_response_->complete());
+  EXPECT_THAT(client_response_->headers(), Http::HttpStatusIs("200"));
+  EXPECT_EQ(client_response_->body().size(), response_size);
+  EXPECT_EQ(processor_response_hash_, HashUtil::xxHash64(client_response_->body()));
+}
+
+// Send a large HTTP POST, and expect back an equally large reply. Stream both and verify
+// that we got back what we expected.
+TEST_P(StreamingIntegrationTest, PostAndProcessStreamBothBodies) {
+  const uint32_t send_chunks = 10;
+  const uint32_t chunk_size = 11000;
+  uint32_t request_size = send_chunks * chunk_size;
+  uint32_t response_size = 1700000;
+
+  test_processor_.start(
```

Comment:
These gRPC servers are kind of large, hard to read and their logic is not inline with the rest of the test. Is there a reason we went this path vs. what is done in the rest of the e2e integation tests in Envoy? Can more boiler plate be factored out? Can they be made more declarative?

Reply:
"streaming_integration_test" is a new test I introduced in the last PR. Those other tests are still there, in "ext_proc_integration_test." 

These tests use a real gRPC server rather than simulating it through the integration framework, and it runs in a separate thread. To test streaming I wanted to be able to stream chunks of the body in both directions and the integration framework seems more designed to do things in a linear way -- send body from downstream, read it upstream, send it back upstream, read the result -- and in reality all those things may happen at once.

I personally find this "grpc server" style much easier for me to read and write because the ext_proc server that we're testing is the same server that a user would actually create, rather than a simulation, and IMO it's closer to replicating the behavior of the real thing in which chunks of data are passing through the filter, to the server, and back again, all in parallel.

Reply:
But yes, they could be shrunk with some helper functions.

Reply:
I think at least reducing boiler plate would be good. It's hard to follow the test really as it's currently written.

Reply:
Correct me if I'm wrong. My understanding is the gRPC server is necessary because it introduces different interleaving of headers, body chunks, etc. and the functions provided in the HTTPIntegration class are "all or nothing" sending a full request and waiting for a full response. The downside of this approach is that we cannot recreate a specific interleaving if it causes an issue.

If the above is true, this is extremely related to my fuzzing goal. My goal is to be able to control how many chunks are sent, when they are sent, when the mock gRPC server responds to the filter, the number of chunks the upstream responds with, and when the upstream responds. Although, I'm not sure if the filter is deterministic when it has both data from the downstream and upstream waiting to be processed.

Is this level of granularity something that would be useful for generic ext_proc integration tests, @gbrail? One downside that immediately comes to mind is that writing tests where you manually specific the downstream, upstream, and gRPC server actions will be very tedious and possibly result in long test cases.

Reply:
I think that's a fair description; you lose thread interleaving determinism unless the RPC sequences effectively serialize this, which I don't think is true with the `ext_proc` async RPC design in general.

Reply:
I made the tests a bit better with a helper function and by switching to standard GTest macros. For this PR I added both these "streaming integration tests" and also new mock-based tests with a deterministic order in filter_test.cc, so I think that we're fairly well covered, but I'm also looking forward to the fuzz tests.